### PR TITLE
New Swing Main Menu (fixing #268)

### PR DESCRIPTION
**Pull request by andreas-eberle**
This is the implementation of #268.

The code should be feature complete and will now be reviewed. Please try out the new UI and help to find bugs. Plese note the following limitations.

**Limitations**
* Multiplayer does not support many of the features you can configure in single player
* The game logic cannot handle peace time and teams yet. This will be implemented later.

### CODE REVIEW

File: jsettlers.main.swing/src/jsettlers/main/swing/lookandfeel/ui/ScrollbarUi.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		incrGap = 0;
+		decrGap = 0;
+	}
+
+	@Override
+	public Dimension getPreferredSize(JComponent c) {
+		if (scrollbar.getOrientation() == JScrollBar.VERTICAL) {
+			return new Dimension(scrollBarWidth, scrollBarWidth * 3 + 10);
+		} else { // Horizontal
+			return new Dimension(scrollBarWidth * 3 + 10, scrollBarWidth);
```
Comment (andreas-eberle) [inline jsettlers.main.swing/src/jsettlers/main/swing/lookandfeel/ui/ScrollbarUi.java]: @andreasb242, @codingberlin 
Is it desired to set a fixed aspect ratio here?

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/general/VolumeSlider.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/**
+ * Slider to select volume in settings
+ * 
+ * @author Andreas Butti
+ */
+public class VolumeSlider extends SettlerSlider {
+	private static final long serialVersionUID = 1L;
+
+	/**
+	 * Constructor
```
Comment (michaelzangl) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/general/VolumeSlider.java]: Could you please add a more verbose javadoc comment? You could e.g. name the default values used.

File: jsettlers.common/src/jsettlers/common/ai/EPlayerType.java [hunk 3]
```diff
@@ -17,16 +17,21 @@
...
+public enum EPlayerType {
+	AI_VERY_EASY,
+	AI_EASY,
+	AI_HARD,
+	AI_VERY_HARD,
+	HUMAN;
 
-	public static final EWhatToDoAiType[] values = EWhatToDoAiType.values();
-	public static final int NUMBER_OF_AI_TYPES = values.length;
+	public static final EPlayerType[] values = EPlayerType.values();
```
Comment (michaelzangl) [inline jsettlers.common/src/jsettlers/common/ai/EPlayerType.java]: Can you change this to VALUES?

File: jsettlers.main.swing/src/jsettlers/main/swing/lookandfeel/ui/ScrollbarUi.java (hunk 1 above)
Comment (andreasb242) [inline jsettlers.main.swing/src/jsettlers/main/swing/lookandfeel/ui/ScrollbarUi.java]: The Scrollbar itself is scaleable, so this is the minimum size which is needed to display the scrollbar. In BasicScrollbarUI the width is hardcoded, which didn't match the graphics I drawed.

Even if there is also a minimum and maximum size, this is usually not working with the most layout manager...

So no, this does not set a fixed aspect ration.

Comment (andreas-eberle) [inline jsettlers.main.swing/src/jsettlers/main/swing/lookandfeel/ui/ScrollbarUi.java]: ok.

File: jsettlers.main.swing/src/jsettlers/main/swing/menu/general/VolumeSlider.java (hunk 2 above)
Comment (andreasb242) [inline jsettlers.main.swing/src/jsettlers/main/swing/menu/general/VolumeSlider.java]: @codingberlin 
It's your pull request, can you may add:

> This slider is technically based on a progress bar, but looks and works like the production
> sliders in the original game. (blue bars)
> The slider lets the user select a value from 0 to 100%, the value is also displayed as string

**Comment by andreas-eberle** [comment]
@andreasb242: Is there a real difference in the functionality of SettlersLabelUi and SettlersDynamicLabelUi? Or can one be replaced by the other? If not, is it possible that one of them inherites the other?

**Comment by andreasb242** [comment]
@andreas-eberle "SettlersLabelUi" has a fixed size, and use an original image as background. But it was to short for some Labels, so I created additional SettlersDynamicLabelUi, where I repeat a part of the image, is working but doesn’t look that good. (I think this UI was only used for one label?)
inherites makes no sense.

I'm still waiting for the merge, I would like to create a full dynamic label, which replace *both*:
* First drawing a background texture, which is repeated
* Draw a scaled half transparent image over this, to get a "gradient" (Exactly what I did for the main background)
* Draw the label border
* And the shadowed Text

This UI should have a dynamic size but no visible image repeat, but it's hard to create good looking texture, a graphics designer would be nice ;-)

**Comment by abledbody** [comment]
@andreasb242 You're in luck, buddy! I could easily help with something like that.

**Comment by andreasb242** [comment]
@andreas-eberle leafe this implementation for the moment, and fix it with Ticket #332
@abledbody I created a Ticket, I could not find any contact information of you, so I published my Email address in my profile.
I mentioned you in the Ticket, you may have a look at #332
You may write me a mail. ps. next week I'm away and will not respond to mails...

**Comment by michaelzangl** [comment]
@andreas-eberle config.prp ;-)

**Comment by andreas-eberle** [comment]
I just finished going through the code and will now merge the PR.

Thanks a lot to @codingberlin and @andreasb242 for accomplishing this great new main menu! This is really a big step forward for our remake! 

I'm looking forward to your further contributions! ;)

**Pull request merged**